src/server/admin/cmds/cmds.go

```diff
@@ -32,20 +34,46 @@ $ {{alias}} -u s3://bucket/backup`,
 			if err != nil {
 				return err
 			}
+
+			// If the cluster has auth enabled, mint a robot token as a backup
+			// in case auth is misconfigured on restore.
+			authActive, err := c.IsAuthActive()
+			if err != nil {
+				return errors.Wrapf(grpcutil.ScrubGRPC(err), "error")
+			}
+
+			if authActive {
+				resp, err := c.GetAuthToken(c.Ctx(), &auth.GetAuthTokenRequest{
```

Comment:
Because of the way the API is structured, the only place to generate a new token and show the user the plaintext is in the CLI itself. Only the hashed robot token is stored in the extract.

Reply:
I don't totally understand the big picture here. It looks like the server-side `Extract()` implementation generates an `Activate` message if auth is active, which I'd expect to create the first admin in the target cluster during `Restore()`, so I don't totally understand what the user is supposed to do with this token after we print it? Is `resp` used for other stuff in this function and I just missed it?

Reply:
`Activate` creates the first admin in the restored cluster, but how do we plumb that information (the plaintext for the first auth token) back to the user doing the restore? The `Restore` RPC returns `Empty`, which seems like an oversight. We could change that to a `RestoreResponse`, and try and put the auth token in there - not sure if this counts as a breaking API change.

My other concern with putting the only known-good auth token in the RPC response to `Restore` is that, if something later in the restore blows up, that response won't be sent to the user, and the cluster may be totally useless. If we generate a token for the user as part of the Extract process, as long as Activate and restoring the auth tokens succeeds, the user has all their data, pipelines and an auth identity they could use to manually clean up the mess.

A third option I considered was having `Activate` accept a hashed auth token to use for the initial admin. In this scenario the hashed root token would be baked into the extract, and the user restoring the extract would supply the plaintext for the token in the auth context from pachctl. This has the nice property that it doesn't generate any new tokens during the restore process.

Reply:
Ah, now I see the issue, thanks! One question: above, you mentioned that you wanted to make `RootUser` the hard-coded default robot account in 2.0 (meaning they're always a super admin?). Will pach 1.12 will work that way? I ask because my understanding of the mechanism here is: any token for `RootUser` is an escape hatch, and this gives users an escape hatch by getting a new `RootUser` token and just printing it, and I want to make sure that's right. If that's how this works, does `Activate()` guarantee that `RootUser` is a cluster admin in 1.12? This approach broadly makes sense to me, and I'm on board with it, though I outlined a few alternatives that occurred to me below as well, in case they're appealing.

- One high-level option that occurs to me is that we could make `Restore()` a streaming API (or add a new `RestoreAuthCompatible()` API that is streaming, as an alternative to the existing API, per semver. If someone calls the old API and it encounters an `Activate` op, or we use the no-op auth marker described below and the old API encounters that, it could simply error). This would let `Restore()` generate a token if it encounters the no-op auth marker at the beginning and send it back right away, but then keep going with the restore. A benefit of this approach is that it avoids user-generated tokens (as described below).

- I do also like the idea of setting up auth using a token the user already has during `Restore()`. Is the idea that the user would send the hash, and we'd call `Activate(hash) -> tokens.Put(hash, 'robot:extractrestore')` and then auth is activated and the user's token authenticates them as `robot:extractrestore`?
  - This is a small modification, but if that understanding is right, it seems like one risk with using the user's current token during `Restore()` is that if we activate auth and place `(user_token_hash, robot:extractrestore)` in the tokens collection, and then later call `RestoreToken(user_token_hash, github:alice)` and `alice` isn't an admin (e.g. different user restoring than extracting) then the user might be downgraded from admin to non-admin partway through the restore, which might mess up subsequent restore ops. One way we could work around that is by generating a token client-side before calling `Restore()` and printing _that_ out at the start of `Restore()`. Restore would use the client-generated temporary token as the initial admin token (`robot:extractrestore`), and use that for everything. Another benefit of that approach is that when `Restore()` puts `(client_generated_token_hash, robot:extractrestore)` in the tokens collection, we could put e.g. a 30-day TTL on it, which prevents the token-leaking issue. Alternatively, `Restore()` could put `(user_generated_token_hash, RootUser)` in the tokens collection with no lease?
  - A closely-related corner case is that if the user doesn't have an auth token in their config at all, but you encounter an `Activate()` request partway through a big restore, then you have to send the user an error after already restoring a bunch of stuff. But one solution that occurs to me is to have the first message in the Restore dump be a no-op marker that just indicates "auth will be activated later, watch out," and if `Restore()` sees that but the user didn't send `Restore()` a token, we can error right away (also, encountering an `Activate()` request if there was no auth marker at the beginning would be an error, but that would be a bug in `Extract()` rather than user error).
  - One final idea: on an API level, as far as activating auth with a user-provided token hash, we could lean on `Activate()`, or we could modify `RestoreToken()` to activate auth if it's not already activated. The internal implementation could be roughly the same—Activate and RestoreToken could use a shared helper—but it would restrict the use of client-generated tokens to a less-used API.

Reply:
OK, I've taken another swing at this:

- Activate now accepts an optional token hash to use as the root token. I realized I want this for integration tests anyways, so we don't have to deal with randomly generated tokens. 

- If auth is activated on the cluster, the client generates a token and includes it in the Extract request.

- Extract includes a `CheckAuthToken` op in the extract file which verifies the auth token during Restore.

- Once Extract has succeeded the client spits out the token on stdout.

- The user must set up their context for Restore by activating the token ahead of time.

- The user calls Restore and the auth token is included in the GRPC context.

- Server-side the Restore method immediately sees the `CheckAuthToken` op and verifies that the token in the context matches the hashed token in the extract.

- If the tokens match, the server will call Activate with the hashed token. This atomically enables auth, sets the root token and makes RootUser an admin. This way the cluster is never in an unrecoverable state.

> Ah, now I see the issue, thanks! One question: above, you mentioned that you wanted to make RootUser the hard-coded default robot account in 2.0 (meaning they're always a super admin?). Will pach 1.12 will work that way? 

I think this is too much of a breaking change for a minor release. In 2.0 I want to take away the ability to remove RootUser from the set of admins, and always use that account for Activate.

> If that's how this works, does Activate() guarantee that RootUser is a cluster admin in 1.12?

I realized the previous implementation didn't guarantee this. The new one does. There's nothing that happens after Activate that removes admins, Restore only adds new ones.

> One high-level option that occurs to me is that we could make Restore() a streaming API (or add a new RestoreAuthCompatible() API that is streaming, as an alternative to the existing API, per semver.

In my experience bidirectional streaming grpc APIs are kind of gross to work with. I'm hoping we can come up with a solution where we don't need to stream data back to the client mid-Restore.

> One way we could work around that is by generating a token client-side before calling Restore() and printing that out at the start of Restore(). Restore would use the client-generated temporary token as the initial admin token (robot:extractrestore), and use that for everything.

The upside of generating the token in Restore is there's no risk a user will lose their token (and consequently their ability to restore a backup). The downside is that Restore doesn't really know whether or not a given extract includes auth until after it's been extracted. We would either print the token immediately when Restore is called (even if auth is not enabled), print it when we get some streaming message (potential timing issues?) or print it after the restore is successful (half-restored clusters can be in a bad state).

> Another benefit of that approach is that when Restore() puts (client_generated_token_hash, robot:extractrestore) in the tokens collection, we could put e.g. a 30-day TTL on it, which prevents the token-leaking issue. Alternatively, Restore() could put (user_generated_token_hash, RootUser) in the tokens collection with no lease?

I would prefer not to put a TTL on the only admin token we know a user has, as it would basically make the cluster a time bomb. I guess they could always restore again, but if they did anything in the meantime they could end up in a bad state. In the future if we're concerned about the token leaking issue we can avoid restoring any other RootUser tokens _except_ the one printed by Extract.